subprojects/plugins/src/main/java/org/gradle/api/plugins/internal/DefaultJavaPluginExtension.java

```diff
@@ -66,7 +66,12 @@ public DefaultJavaPluginExtension(JavaPluginConvention convention,
         this.project = project;
         this.modularity = objectFactory.newInstance(DefaultModularitySpec.class);
         this.jvmPluginServices = jvmPluginServices;
-        this.toolchain = objectFactory.newInstance(DefaultToolchainSpec.class);
+        DefaultToolchainSpec toolchainSpec = objectFactory.newInstance(DefaultToolchainSpec.class);
+        this.toolchain = toolchainSpec;
+        if (convention instanceof DefaultJavaPluginConvention) {
+            DefaultJavaPluginConvention javaPluginConvention = (DefaultJavaPluginConvention) convention;
+            javaPluginConvention.setToolchainSpec(toolchainSpec);
```

Comment:
Why did you choose to do this after the fact and make the convention mutable? I think we could instantiate the convention with the spec in the [base plugin](https://github.com/gradle/gradle/blob/master/subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java#L136) and share that same spec with the extension.

Reply:
🤦 of course, did not think of that

Reply:
And no need for the cast then!